Swift/Tests/PredictiveQueryTest+CoreML.swift

```diff
@@ -0,0 +1,221 @@
+        let prediction = Function.prediction(model: "MobileNet", input: input)
+        let q = QueryBuilder
+            .select(SelectResult.expression(prediction))
+            .from(DataSource.database(db))
+        let rows = try verifyQuery(q) { (n, r) in
+            let pred = r.dictionary(at: 0)!
+            let label = pred.string(forKey: "classLabel")!.lowercased()
+            XCTAssertNotEqual((label as NSString).range(of: "cat").location, NSNotFound)
+            let probs = pred.dictionary(forKey: "classLabelProbs")!
+            XCTAssertTrue(probs.count > 0)
+        }
+        XCTAssertEqual(rows, 1);
+        
+        Database.prediction.unregisterModel(withName: "MobileNet")
+    }
+    
+    // Note: Download OpenFace.mlmodel from https://github.com/iwantooxxoox/Keras-OpenFace
+    // and put it at Objective-C/Tests/Support/mlmodels/OpenFace
+    func testOpenFaceModel() throws {
+        guard let model = try self.model(name: "OpenFace/OpenFace", mustExist: false) else {
+            return
+        }
+        Database.prediction.registerModel(model, withName: "OpenFace")
+        
+        let faces = ["adams", "lennon-3", "carell", "lennon-2", "lennon-1"]
+        for face in faces {
+            try crateDocumentWithImage(at: "mlmodels/OpenFace/\(face).png")
+        }
+        
+        // Query the finger print of each face:
+        var result: [String: ArrayObject] = [:]
+        let input = Expression.dictionary(["data" : Expression.property("image")])
+        let prediction = Function.prediction(model: "OpenFace", input: input)
+        var q: Query = QueryBuilder
+            .select(SelectResult.property("name"), SelectResult.expression(prediction))
+            .from(DataSource.database(db))
+        var rows = try verifyQuery(q) { (n, r) in
+            let name = r.string(at: 0)!
+            let pred = r.dictionary(at: 1)!
+            let output = pred.array(forKey: "output")!
+            XCTAssertEqual(output.count, 128)
+            result[name] = output
+        }
+        XCTAssertEqual(rows, UInt64(faces.count));
+        
+        // Query the euclidean distance between each face and lennon-1:
+        var names: [String] = []
+        let lennon1 = result["lennon-1"]
+        let vector1 = Expression.parameter("vectorParam")
+        let vector2 = prediction.property("output")
+        let distance = Function.euclideanDistance(between: vector1, and: vector2)
+        q = QueryBuilder
+            .select(SelectResult.property("name"), SelectResult.expression(distance))
+            .from(DataSource.database(db))
+            .orderBy(Ordering.expression(distance))
+        let params = Parameters()
+        params.setArray(lennon1, forName: "vectorParam")
+        q.parameters = params
+        rows = try verifyQuery(q) { (n, r) in
+            let name = r.string(at: 0)!
+            names.append(name)
+            XCTAssertNotNil(r.value(at: 1))
+            XCTAssertTrue(r.double(at: 1) >= 0)
```

Comment:
can the distance anytime become less than zero?

Reply:
No, based on https://en.wikipedia.org/wiki/Euclidean_distance.

Reply:
can we validate the `lennon-1` distance is 0? because they are equal right.

Reply:
Yes, we can add that check.